src/main/scala/com/github/jberggg/router/Service.scala

```diff
@@ -0,0 +1,42 @@
+package com.github.jberggg.router
+
+import cats.implicits._
+import org.http4s.Uri.Path
+
+import org.scalajs.dom.window
+import cats.effect.std.Dispatcher
+import org.scalajs.dom.Event
+import fs2.concurrent.Channel
+import fs2._
+import cats.effect.kernel.{Async, Resource}
+
+object Service {
+  
+    def setupInfrastructure[F[_] : Async ]: F[Channel[F, Path]] = for {
+        channel     <- Channel.unbounded[F,Path]
+        location    <- getCurrentLocation[F]
+        _           <- channel.send(location)
+    } yield channel
+
+    private[router] def registerEventHandlerAndToStream[ F[_] : Async ](c: Channel[F, Path]): Stream[F, Path] = for {
+        dispatcher <- Stream.resource(Dispatcher[F])
+        callback   =  (_: Event) => dispatcher.unsafeRunAndForget( getCurrentLocation[F].flatMap( l => c.send(l) ) )
+        _          <- Stream.resource(Resource.make(addHashChangeListener[F](callback))(_ => removeHashChangeListener[F](callback)))
+        paths      <- c.stream
+    } yield paths
+
+    private def getCurrentLocation[F[_] : Async]: F[Path] = Async[F].delay(Path.unsafeFromString(window.location.hash.tail))
+
+    private def addHashChangeListener[ F[_] : Async ](callback: Event => Unit): F[Unit] =  
+        Async[F].delay( window.addEventListener(
+            `type` = "hashchange", 
+            listener = callback,
+            useCapture = true
+        ))
+
+    private def removeHashChangeListener[ F[_] : Async ](originalCallback: Event => Unit): F[Unit] =
+        Async[F].delay( window.removeEventListener(
+            `type` = "hashchange", 
+            listener = originalCallback
+        ))
```

Comment:
Ah, will have a look into that, thank you for the input 👍🏽

Reply:
I changed this logic to ensure that the callback is a `js.Function`. I also tried to model it as an Resource but this somehow didn't work as expected. The callback was not properly registered. Not totally sure why and I ran out of time to be honest :)

Reply:
Yup, it's definitely much better! Up to you, here's an example that uses an `AbortController` (instead of the `js.Function` trick) to create a `Resource`.
https://github.com/armanbilge/calico/blob/7d0b46e56a9cfb04556843417abb8155ac4d6d3d/calico/src/main/scala/calico/dsl.scala#L288-L299